// vite-plugin-ssr/dist/cjs/node/plugin/plugins/config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.setVitePluginSsrConfig = void 0;
const prerenderConfig_1 = require("../../prerender/prerenderConfig");
const utils_1 = require("../utils");
const pageFilesConfig_1 = require("./generateImportGlobs/pageFilesConfig");
function setVitePluginSsrConfig(vitePluginSsrConfig = {}) {
    (0, utils_1.assertUsage)((0, utils_1.isObject)(vitePluginSsrConfig), '[vite.config.js][`ssr(options)`] `options` should be an object');
    return {
        name: 'vite-plugin-ssr:setVitePluginSsrConfig',
        enforce: 'pre',
        config() {
            const config = {
                vitePluginSsr: {
                    prerender: (0, prerenderConfig_1.getPrerenderConfig)(vitePluginSsrConfig),
                    pageFiles: (0, pageFilesConfig_1.getPageFilesConfig)(vitePluginSsrConfig),
                },
            };
            return config;
        },
    };
}
exports.setVitePluginSsrConfig = setVitePluginSsrConfig;
