// vite-plugin-ssr/dist/cjs/node/plugin/plugins/extractExportNamesPlugin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.extractExportNamesPlugin = void 0;
const utils_1 = require("../utils");
const parseEsModules_1 = require("../parseEsModules");
const extractExportNamesRE = /(\?|&)extractExportNames(?:&|$)/;
function extractExportNamesPlugin() {
    return {
        name: 'vite-plugin-ssr:extractExportNames',
        enforce: 'post',
        async transform(src, id, options) {
            const isClientSide = !(0, utils_1.isSSR_options)(options);
            if (extractExportNamesRE.test(id)) {
                const code = await getExtractExportNamesCode(src, isClientSide);
                return code;
            }
        },
    };
}
exports.extractExportNamesPlugin = extractExportNamesPlugin;
async function getExtractExportNamesCode(src, isClientSide) {
    const esModules = await (0, parseEsModules_1.parseEsModules)(src);
    const exportNames = (0, parseEsModules_1.getExportNames)(esModules);
    const code = getCode(exportNames, isClientSide);
    return (0, utils_1.removeSourceMap)(code);
}
function getCode(exportNames, isClientSide) {
    let code = '';
    code += '\n';
    code += `export const exportNames = [${exportNames.map((n) => JSON.stringify(n)).join(', ')}];`;
    if (isClientSide) {
        code += '\n';
        code += `if (import.meta.hot) import.meta.hot.accept((mod) => { exportNames.length=0; exportNames.push(...mod.exportNames); });`;
    }
    return code;
}
