// vite-plugin-ssr/dist/esm/client/page-files/pageFiles-client.js
// This file was generatead by `node/plugin/plugins/generateImportGlobs.ts`.

export const pageFilesLazy = {};
export const pageFilesEager = {};
export const pageFilesExportNamesLazy = {};
export const pageFilesExportNamesEager = {};
export const neverLoaded = {};
export const isGeneratedFile = true;

const pageFilesLazyIsomorph1 = import.meta.importGlob('/**/*.page.*([a-zA-Z0-9])', { eager: false, query: "" });
const pageFilesLazyIsomorph2 = import.meta.importGlob('/node_modules/vilay/renderer/**/*.page.*([a-zA-Z0-9])', { eager: false, query: "" });
const pageFilesLazyIsomorph = {...pageFilesLazyIsomorph1,...pageFilesLazyIsomorph2};
pageFilesLazy['.page'] = pageFilesLazyIsomorph;

const pageFilesLazyRoute1 = import.meta.importGlob('/**/*.page.route.*([a-zA-Z0-9])', { eager: false, query: "" });
const pageFilesLazyRoute2 = import.meta.importGlob('/node_modules/vilay/renderer/**/*.page.route.*([a-zA-Z0-9])', { eager: false, query: "" });
const pageFilesLazyRoute = {...pageFilesLazyRoute1,...pageFilesLazyRoute2};
pageFilesLazy['.page.route'] = pageFilesLazyRoute;

const pageFilesLazyClient1 = import.meta.importGlob('/**/*.page.client.*([a-zA-Z0-9])', { eager: false, query: "" });
const pageFilesLazyClient2 = import.meta.importGlob('/node_modules/vilay/renderer/**/*.page.client.*([a-zA-Z0-9])', { eager: false, query: "" });
const pageFilesLazyClient = {...pageFilesLazyClient1,...pageFilesLazyClient2};
pageFilesLazy['.page.client'] = pageFilesLazyClient;

const pageFilesExportNamesLazyClient1 = import.meta.importGlob('/**/*.page.client.*([a-zA-Z0-9])', { eager: false, query: "extractExportNames" });
const pageFilesExportNamesLazyClient2 = import.meta.importGlob('/node_modules/vilay/renderer/**/*.page.client.*([a-zA-Z0-9])', { eager: false, query: "extractExportNames" });
const pageFilesExportNamesLazyClient = {...pageFilesExportNamesLazyClient1,...pageFilesExportNamesLazyClient2};
pageFilesExportNamesLazy['.page.client'] = pageFilesExportNamesLazyClient;

const pageFilesExportNamesLazyServer1 = import.meta.importGlob('/**/*.page.server.*([a-zA-Z0-9])', { eager: false, query: "extractExportNames" });
const pageFilesExportNamesLazyServer2 = import.meta.importGlob('/node_modules/vilay/renderer/**/*.page.server.*([a-zA-Z0-9])', { eager: false, query: "extractExportNames" });
const pageFilesExportNamesLazyServer = {...pageFilesExportNamesLazyServer1,...pageFilesExportNamesLazyServer2};
pageFilesExportNamesLazy['.page.server'] = pageFilesExportNamesLazyServer;

const pageFilesExportNamesLazyIsomorph1 = import.meta.importGlob('/**/*.page.*([a-zA-Z0-9])', { eager: false, query: "extractExportNames" });
const pageFilesExportNamesLazyIsomorph2 = import.meta.importGlob('/node_modules/vilay/renderer/**/*.page.*([a-zA-Z0-9])', { eager: false, query: "extractExportNames" });
const pageFilesExportNamesLazyIsomorph = {...pageFilesExportNamesLazyIsomorph1,...pageFilesExportNamesLazyIsomorph2};
pageFilesExportNamesLazy['.page'] = pageFilesExportNamesLazyIsomorph;

const neverLoadedServer1 = import.meta.importGlob('/**/*.page.server.*([a-zA-Z0-9])', { eager: false, query: "extractStyles" });
const neverLoadedServer2 = import.meta.importGlob('/node_modules/vilay/renderer/**/*.page.server.*([a-zA-Z0-9])', { eager: false, query: "extractStyles" });
const neverLoadedServer = {...neverLoadedServer1,...neverLoadedServer2};
neverLoaded['.page.server'] = neverLoadedServer;
