// vite-plugin-ssr/dist/esm/node/page-files/pageFiles-node.js
// This file was generatead by `node/plugin/plugins/generateImportGlobs.ts`.

export const pageFilesLazy = {};
export const pageFilesEager = {};
export const pageFilesExportNamesLazy = {};
export const pageFilesExportNamesEager = {};
export const neverLoaded = {};
export const isGeneratedFile = true;

const pageFilesLazyIsomorph1 = import.meta.importGlob('/**/*.page.*([a-zA-Z0-9])', { eager: false, query: "" });
const pageFilesLazyIsomorph2 = import.meta.importGlob('/node_modules/vilay/renderer/**/*.page.*([a-zA-Z0-9])', { eager: false, query: "" });
const pageFilesLazyIsomorph = {...pageFilesLazyIsomorph1,...pageFilesLazyIsomorph2};
pageFilesLazy['.page'] = pageFilesLazyIsomorph;

const pageFilesLazyRoute1 = import.meta.importGlob('/**/*.page.route.*([a-zA-Z0-9])', { eager: false, query: "" });
const pageFilesLazyRoute2 = import.meta.importGlob('/node_modules/vilay/renderer/**/*.page.route.*([a-zA-Z0-9])', { eager: false, query: "" });
const pageFilesLazyRoute = {...pageFilesLazyRoute1,...pageFilesLazyRoute2};
pageFilesLazy['.page.route'] = pageFilesLazyRoute;

const pageFilesLazyServer1 = import.meta.importGlob('/**/*.page.server.*([a-zA-Z0-9])', { eager: false, query: "" });
const pageFilesLazyServer2 = import.meta.importGlob('/node_modules/vilay/renderer/**/*.page.server.*([a-zA-Z0-9])', { eager: false, query: "" });
const pageFilesLazyServer = {...pageFilesLazyServer1,...pageFilesLazyServer2};
pageFilesLazy['.page.server'] = pageFilesLazyServer;

const pageFilesExportNamesLazyClient1 = import.meta.importGlob('/**/*.page.client.*([a-zA-Z0-9])', { eager: false, query: "extractExportNames" });
const pageFilesExportNamesLazyClient2 = import.meta.importGlob('/node_modules/vilay/renderer/**/*.page.client.*([a-zA-Z0-9])', { eager: false, query: "extractExportNames" });
const pageFilesExportNamesLazyClient = {...pageFilesExportNamesLazyClient1,...pageFilesExportNamesLazyClient2};
pageFilesExportNamesLazy['.page.client'] = pageFilesExportNamesLazyClient;
